BinaryTree.java

```diff
@@ -1,6 +1,15 @@
+import java.util.ArrayList;
+import java.util.List;
+
 public class BinaryTree<T> {
+    private final BinaryTreeNode<T> root;
+    public enum InsertChild {Left, Right};
+
+    private InsertChild m_insertChild = InsertChild.Left;

```

Comment:
Is storing direction of child insertion inside binary tree really needed? Having container in a state that would only insert to the left or only to the right does not seem too useful and would be a pain to use.

Why not just passing this enum as parameter to insert method?